### DISCUSSION TITLE
Have a `Phlex::Text` to build large plain texts?

### DISCUSSION THREAD

**Discussion by fnordfish** [yippee-fun/phlex#829] [Ideas]
I'm trying to convert a bunch of string templates used for generating email texts to Phlex.
So I was wondering if we could have a `Text` or, `Plain`, or `PlainText` base class, that simply emits text without any HTML escaping but still supports layouts and components/kits.
Maybe even have build in helpers for line/word wrapping.

I guess it could be hacked by using `Phlex::SGML` and use `raw(safe("text"))` on everything.

**Comment by joeldrapper** [comment]
Would you not want HTML escaping anyway, even if the output was meant to be plain text? In email, for example, you wouldn’t want to unintentionally allow unsafe user input to produce HTML.

I think it would help if we had an example use-case. Could you write an example of what you'd want `Phlex::Plain` to be like in terms of source code and output?

**Reply by fnordfish** [reply]
Right, I wouldn't want any escaping in `Plain`. In my current use case, the generated text will be fed to the mail gem and properly escaped there. 

The idea is to generate text with re-usable components where the output is actually meant to be text. But also constructing long texts to be re-used in HTML or SVG.

```ruby
class Phlex::Plain < Phlex::SGML
	# Just for the hacky POC
	def line(str = "", ending: "\n")
		raw(safe(str + ending))
	end
end

module Components
	extend Phlex::Kit

	class PlainLayout < Phlex::Plain
		def initialize(name = "World")
			@name = name
		end

		def view_template(&)
			line("Hello #{@name}!")
			raw(yield)
			line("Bye!")
		end
	end
end

class Example < Phlex::Plain
	include Components

	def initialize(items)
		@items = items
	end

	def view_template
		Components::PlainLayout() do
			line("Here are some items:")
			@items.each do |item|
				line("- #{item}")
			end
		end
	end
end

class ExampleHtml < Phlex::HTML
	include Components

	def initialize(items)
		@items = items
	end

	def view_template
		pre {
			Example.new(@items).call
		}
	end
end

test "render plain text with layout" do
	text = Example.new(["Zöe's Café & Bakery", "<Acme Super Cleaner>", "Gemüse"]).call
	expect(text) == <<~PLAIN
	Hello World!
	Here are some items:
	- Zöe's Café & Bakery
	- <Acme Super Cleaner>
	- Gemüse
	Bye!
	PLAIN
end

test "render plain text inside HTML" do
	html = ExampleHtml.new(["Zöe's Café & Bakery", "<Acme Super Cleaner>", "Gemüse"]).call
	expect(html) == <<~HTML.chomp
	<pre>Hello World!
	Here are some items:
	- Zöe&#39;s Café &amp; Bakery
	- &lt;Acme Super Cleaner&gt;
	- Gemüse
	Bye!
	</pre>
	HTML
end
```
